include/seastar/websocket/client.hh

```diff
@@ -0,0 +1,140 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <map>
+#include <functional>
+#include <random>
+
+#include <seastar/core/sstring.hh>
+#include <seastar/http/request.hh>
+#include <seastar/http/connection_factory.hh>
+#include <seastar/websocket/common.hh>
+
+namespace seastar::experimental::websocket {
+
+class client;
+
+/// \addtogroup websocket
+/// @{
+
+/*!
+ * \brief a client WebSocket connection
+ */
+class client_connection : public connection {
+    client& _client;
+    sstring _ws_key;
+public:
+    /*!
+     * \param server owning \ref server
+     * \param fd established socket used for communication
+     */
+    client_connection(client& client, connected_socket&& fd, std::string_view ws_key,
+                      handler_t handler);
+    ~client_connection();
+
+    /*!
+     * \brief serve WebSocket protocol on a client_connection
+     */
+    future<> process();
+
+    /**
+     * @brief Send a websocket message to the server
+     */
+    future<> send_message(temporary_buffer<char> buf, bool flush);
+
+protected:
```

Comment:
Why no private? Are there plans to implement more specific connections inherited from this class?